[master] Enhancing the ps beacon.

KiyoIchikawa opened the pull request:
Enhancing the ps beacon to return username of the process and the create_time. The ps beacon can also check for processes running under a specific user. Updated doc strings and added more unit tests.

### What does this PR do?
* ps Beacon will now:
  * Return username of the process and the create_time.
  * Returns multiple instances of process to watch if a username is not specified in the config.
  * Have the ability to check for a process running under a specific user.
  * Beacon now returns empty list if expected status of process not found.
  * Restricted to checking for specific process statuses and can be extended later.
* Add unit tests to support new functionality as well as scenarios not previously covered.
* Updated doc strings to reflect new behavior.

### What issues does this PR fix or reference?
Fixes: N/A

### Previous Behavior
Beacon would return a list of objects with a key of the process name and the discovered status as a string "Running" or "Stopped". If the process to watch is not found the returned dictionary with the process name as the key and False as the value. Example config and return values:

```
# config:
beacons:
  ps:
    - processes:
       process_name: running
       other_process: stopped
       some_other: oop
[
  {
    "process_name": "Running" <-- Found the process is running, expected state.
  },
  {
    "other_process": "Stopped" <-- Found the process is not running, expected state.
  },
  {
    "some_other": False <-- Process not found and config provided invalid status to check for.
  }
]
```

### New Behavior
* Return object now can return multiple instances of watched process if running under multiple users as well as the create_time and PID. If expected statuses are not found, empty list is returned. Example return values and config:
```
config:

beacons:
  ps:
    processes:
      - process_name:
         status: running
      - other_process:
         status: stopped
      - some_other:
         status: oop <-- Get a config validation error.
[
  {
    "process_name": {
      "status": "running",
      "instances": [
        (24, "kichikawa", 1307289803.47),
        (45, "tuser", 1307289703.47)
      ]
    },
    {
      "other_process": {
        "status": "stopped",
        "instances": []
    }
  }
]
```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

KiyoIchikawa commented:
This fixes #65921

KiyoIchikawa commented:
Working to resolve CI failures.

twangboy commented on salt/beacons/ps.py:
```diff
@@ -31,62 +50,85 @@ def validate(config):
     Validate the beacon configuration
     """
     # Configuration for ps beacon should be a list of dicts
```
This comment needs to change too

twangboy commented on salt/beacons/ps.py:
```diff
@@ -31,62 +50,85 @@ def validate(config):
         else:
-            if not isinstance(config["processes"], dict):
+            if not isinstance(config["processes"], list):
                 return False, "Processes for ps beacon must be a dictionary."
```
We're checking for a list, but warning that it must be a dictionary...

twangboy commented on salt/beacons/ps.py:
```diff
@@ -31,62 +50,85 @@ def validate(config):
-
-    The config above sets up beacons to check that
-    processes are running or stopped.
+    Search for given process(es) by name and optionally username.
```
Add back a proper Example Config, especially if you're not going to document the `config` parameter.

twangboy commented on salt/beacons/ps.py:
```diff
@@ -31,62 +50,85 @@ def validate(config):
+    Retrieves process information for the given process.
+    Using x.oneshot() speeds up retrieving multiple attributes for a process and caches the results.
+    More Info: https://psutil.readthedocs.io/en/latest/#psutil.Process.oneshot
+    """
```
Please document the `process` parameter and add a `.. version-added::` tag

twangboy reviewed (changes requested):
It seems that changing the expected format of the beacon config from a list to a dictionary is a breaking change. Please make that explicit in the changelog. We may need to have a discussion around this as well.
Also, this needs to be rebased and conflicts resolved.

bdrx312 commented on salt/beacons/ps.py:
```diff
@@ -32,62 +50,85 @@ def validate(config):
     return True, "Valid beacon configuration"
 
 
 def beacon(config):
```
Some typehints would be really useful.

dwoz committed:
Enhance ps beacon with username, create_time, and user filtering

- Replace psutil_compat import with direct psutil import (psutil_compat
  was removed from all branches in July 2024)
- Config format changed: processes is now a list of dicts with a
  'status' key (and optional 'username') instead of a flat string value
- Add status validation against __accepted_statuses__ in validate()
- Return username and create_time alongside PID in each instance tuple
- Support multiple instances of the same process name
- Support filtering by username when 'username' is set in config
- Add safe wrappers for name()/username() to handle NoSuchProcess etc.
- Replace public context_pull_props() with _context_pull_props()
- Update tests to cover all new scenarios (12 tests, all passing)
- Add changelog/65922.changed.md documenting the breaking config change

dwoz committed:
Address review: add type hints, Example Config, versionchanged, and param docs

- Add type hints to all public and private functions
- Add Example Config section to beacon() docstring
- Add .. versionchanged:: 3009.0 tag documenting the breaking config change
- Document config and return values with :param:/:returns:/:rtype: for
  validate(), beacon(), and _context_pull_props()
